thetis/tracer_eq_2d.py

```diff
@@ -256,20 +307,52 @@ class SourceTerm(TracerTerm):
     """
     def residual(self, solution, solution_old, fields, fields_old, bnd_conditions=None):
         f = 0
+
         source = fields_old.get('source')
+
         if source is not None:
-            f += -inner(source, self.test)*self.dx
+            if self.options.use_tracer_conservative_form:
+                H = self.get_total_depth(fields["elev_2d"])
+                f += -inner(H * source, self.test)*self.dx
+            else:
+                f += -inner(source, self.test)*self.dx
+
         return -f
 
 
+class MassTerm(TracerTerm):
+    r"""
+    Mass term for 2d tracer equation.
+
+    In the default form (non-conservative) this is just the standard mass term
+
+    .. math::
+        F_s = \int_\Omega T \phi
+
+    In the conservative form we multiply by a depth H
+
+    .. math::
+        F_s = \int_\Omega T H \phi
+
+    """
+    def residual(self, solution, fields):
```

Comment:
It's a bit unfortunate that the mass term now needs the fields as an argument. We should discuss this